controllers/managedfusion_controller.go

```diff
@@ -689,6 +723,158 @@ func (r *ManagedFusionReconciler) reconcilePrometheusProxyNetworkPolicy() error
+		if smtpURL.Hostname() == "" {
+			return fmt.Errorf("smtp config does not contain a host entry")
+		}
+
+		smtpEgressRule := ovnv1.EgressFirewallRule{}
+		smtpEgressRule.To.DNSName = smtpURL.Hostname()
+		smtpEgressRule.Type = ovnv1.EgressFirewallRuleAllow
+
+		desired.Spec.Egress = append(
+			[]ovnv1.EgressFirewallRule{
+				smtpEgressRule,
+			},
+			desired.Spec.Egress...,
+		)
+		r.egressFirewall.Spec = desired.Spec
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to update egressFirewall: %v", err)
+	}
+	return nil
+
+}
+
+func (r *ManagedFusionReconciler) reconcileEgressNetworkPolicy() error {
+	if !r.AvailableCRDs[EgressNetworkPolicyCRD] {
+		r.Log.Info("EgressNetworkPolicy CRD not present, skipping reconcile for egress network policy")
+		return nil
+	}
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.egressNetworkPolicy, func() error {
+		if err := r.own(r.egressNetworkPolicy); err != nil {
+			return err
+		}
+		desired := templates.EgressNetworkPolicyTemplate.DeepCopy()
+
+		// Get the aws config map
+		awsConfigMap := &corev1.ConfigMap{}
+		awsConfigMap.Name = utils.IMDSConfigMapName
+		awsConfigMap.Namespace = r.Namespace
+		if err := r.get(awsConfigMap); err != nil {
+			return fmt.Errorf("unable to get AWS ConfigMap: %v", err)
+		}
+
+		// Get the machine cidr
+		vpcCIDR, ok := awsConfigMap.Data[utils.CIDRKey]
+		if !ok {
+			return fmt.Errorf("unable to determine machine CIDR from AWS ConfigMap")
+		}
+		// Allow egress traffic to that cidr range
+		var vpcEgressRules []openshiftv1.EgressNetworkPolicyRule
+		cidrList := strings.Split(vpcCIDR, ";")
+		for _, cidr := range cidrList {
+			if cidr == "" {
+				continue
+			}
+			vpcEgressRules = append(vpcEgressRules, openshiftv1.EgressNetworkPolicyRule{
+				Type: openshiftv1.EgressNetworkPolicyRuleAllow,
+				To: openshiftv1.EgressNetworkPolicyPeer{
+					CIDRSelector: cidr,
+				},
+			})
+		}
+
+		// Inserting the VPC Egress rule in front of all other egress rules.
+		// The order or rules matter
+		// https://docs.openshift.com/container-platform/4.10/networking/openshift_sdn/configuring-egress-firewall.html#policy-rule-order_openshift-sdn-egress-firewall
+		// Inserting this rule in the front ensures it comes before the EgressNetworkPolicyRuleDeny rule.
+		desired.Spec.Egress = append(
+			vpcEgressRules,
+			desired.Spec.Egress...,
+		)
+
+		smtpURL, err := url.Parse(r.smtpConfigData.Endpoint)
+		if err != nil {
+			return fmt.Errorf("unable to parse SMTP Endpoint url: %v", err)
+		}
+		if smtpURL.Hostname() == "" {
+			return fmt.Errorf("smtp config does not contain a host entry")
+		}
+
+		smtpEgressRule := openshiftv1.EgressNetworkPolicyRule{}
+		smtpEgressRule.To.DNSName = smtpURL.Hostname()
+		smtpEgressRule.Type = openshiftv1.EgressNetworkPolicyRuleAllow
+
+		desired.Spec.Egress = append(
+			[]openshiftv1.EgressNetworkPolicyRule{
+				smtpEgressRule,
+			},
+			desired.Spec.Egress...,
+		)
+		r.egressNetworkPolicy.Spec = desired.Spec
+		return nil
```

Comment:
This is the exact same code as the prev func but with slightly different struct names. 
Having this duplication could lead to some issues solved only on one func and not the other. 

Lets have a `getEgressDesiredState` function that returns `*EgressFirewallSpec`. In the first func we just call it and set it and in this one we call it and translate the desired state from `EgressFirewallSpec` to  `EgressNetworkPolicySpec`

Reply:
It will be a lot of type conversion, let me present something

Reply:
Updated